### PR TITLE
[Merged by Bors] - feat(Data/Fin/Tuple): Define `Fin.take` and initial theorems

### PR DESCRIPTION
**Pull request by quangvdao**
This PR defines `Fin.take`, which takes the first `m` elements of an `n`-tuple (from the left). It requires that `m ≤ n`, and returns an `m`-tuple. This is an analogue of `List.take`.

It also provides initial theorems for `Fin.take`. If all goes well, I plan to add more theorems over time, and also define `rtake` to be taking from the right, similar to `List.rtake`.

Context: I'm working on a formalization of interactive proof systems. As part of that process, I need to define the type of message sent in each round, as something like:
```
def MessageType (numRounds : ℕ) := Fin numRounds → Type
def message (numRounds : ℕ) : ∀ i, (MessageType numRounds) i
```
Then I will also need to talk about composition and restriction of interactive protocols. Composition amounts to `Fin.addCases`, but restriction is my new `Fin.take`, and also `Fin.rtake` in the future (taking from the right).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 1]
```diff
@@ -1071,6 +1072,103 @@ theorem contractNth_apply_of_ne (j : Fin (n + 1)) (op : α → α → α) (g : F
 
 end ContractNth
 
+section Take
+
+variable {n : ℕ} {α : Fin n → Sort*}
+
+/-- Take the first `m` elements of an `n`-tuple where `m ≤ n`, returning an `m`-tuple. -/
+def take (v : (i : Fin n) → α i) (m : ℕ) (h : m ≤ n) : (i : Fin m) → α (castLE h i) :=
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Basic.lean:1080]:
```suggestion
def take (m : ℕ) (h : m ≤ n) (v : (i : Fin n) → α i) : (i : Fin m) → α (castLE h i) :=
```
to better match `Fin.init`

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 2]
```diff
@@ -1071,6 +1072,103 @@ theorem contractNth_apply_of_ne (j : Fin (n + 1)) (op : α → α → α) (g : F
...
+  fun i ↦ v (castLE h i)
+
+@[simp]
+theorem take_def (v : (i : Fin n) → α i) (m : ℕ) (h : m ≤ n) (i : Fin m) :
+    (take v m h) i = v (castLE h i) := rfl
+
+@[simp]
+theorem take_zero (v : (i : Fin n) → α i) : take v 0 (Nat.zero_le n) = fun i ↦ elim0 i := by
+  ext i; exact elim0 i
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Basic.lean:1089]: Can you also have `take_succ : take m m.le_succ v = init v`?

**Review by YaelDillies — COMMENTED** [review]
What's your use case for `Fin.take`?

**Comment by quangvdao** [comment]
> What's your use case for `Fin.take`?

I'm working on a formalization of interactive proof systems. As part of that process, I need to define the type of message sent in each round, as something like:
```
def MessageType (numRounds : ℕ) := Fin numRounds → Type
def message (numRounds : ℕ) : ∀ i, (MessageType numRounds) i
```
Then I will also need to talk about composition and restriction of interactive protocols. Composition amounts to `Fin.addCases`, but restriction is my new `Fin.take`, and also `Fin.rtake` in the future (taking from the right).

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 3]
```diff
@@ -1071,6 +1072,110 @@ theorem contractNth_apply_of_ne (j : Fin (n + 1)) (op : α → α → α) (g : F
...
+section Take
+
+variable {n : ℕ} {α : Fin n → Sort*}
+
+/-- Take the first `m` elements of an `n`-tuple where `m ≤ n`, returning an `m`-tuple. -/
+def take (m : ℕ) (h : m ≤ n) (v : (i : Fin n) → α i) : (i : Fin m) → α (castLE h i) :=
+  fun i ↦ v (castLE h i)
+
+@[simp]
+theorem take_def (m : ℕ) (h : m ≤ n) (v : (i : Fin n) → α i) (i : Fin m) :
```
Comment (kim-em) [inline Mathlib/Data/Fin/Tuple/Basic.lean:1084]: `take_apply`?

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 4]
```diff
@@ -1071,6 +1072,110 @@ theorem contractNth_apply_of_ne (j : Fin (n + 1)) (op : α → α → α) (g : F
...
+      simp_all only [coe_castLE, coe_natAdd, coe_subNat, coe_cast, Nat.add_sub_cancel']
+    rw [take, this, append_right]
+
+@[simp]
+theorem take_init {α : Fin (n + 1) → Sort*} (m : ℕ) (h : m ≤ n) (v : (i : Fin (n + 1)) → α i) :
+    take m h (init v) = take m (Nat.le_succ_of_le h) v := by
+  ext i
+  simp only [take, init]
+  congr
+
```
Comment (kim-em) [inline Mathlib/Data/Fin/Tuple/Basic.lean:1176]: Please add theorems showing that `List.ofFn` and `getElem : (l : List α) → (Fin l.length → α)` intertwine `Fin.take` and `List.take`.

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 5]
```diff
@@ -1071,6 +1072,110 @@ theorem contractNth_apply_of_ne (j : Fin (n + 1)) (op : α → α → α) (g : F
 
 end ContractNth
 
+section Take
```
Comment (kim-em) [inline Mathlib/Data/Fin/Tuple/Basic.lean:1075]: This is already in a big file, perhaps move to `Mathlib/Data/Fin/Tuple/Take.lean`.

**Comment by quangvdao** [comment]
@kim-em I have added four theorems that intertwines `Fin.take` and `List.take` with `List.get` and `List.ofFn`. Two of the theorems have to be stated in terms of heterogeneous equality, since the types are only propeq:
```
/-- `Fin.take` intertwines with `List.take` via `List.get`. -/
theorem list_take_get_heq {α : Type*} {m : ℕ} (l : List α) (h : m ≤ l.length) :
    HEq (l.take m).get (take m h l.get) :=
  (Fin.heq_fun_iff (List.length_take_of_le h)).mpr (by simp)

/-- Alternative version of `list_take_get_heq` with `v : Fin n → α` instead of `l : List α`. -/
theorem list_ofFn_take_get_heq {α : Type*} {m : ℕ} (v : Fin n → α) (h : m ≤ n) :
    HEq ((List.ofFn v).take m).get (take m h v) :=
  (Fin.heq_fun_iff (by simp [h])).mpr (by simp)
  ```
  Would appreciate your feedback on theorem statement & naming. Maybe I should get rid of `heq` in the name?

File: Mathlib/Data/Fin/Tuple/Take.lean [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+@[simp]
+theorem take_one {α : Fin (n + 1) → Sort*} (v : (i : Fin (n + 1)) → α i) :
+    take 1 (Nat.le_add_left 1 n) v = (fun i => v (castLE (Nat.le_add_left 1 n) i)) := by
+  ext i
+  simp only [take]
+
+@[simp]
+theorem take_of_succ {α : Fin (n + 1) → Sort*} (v : (i : Fin (n + 1)) → α i) :
+    take n n.le_succ v = init v := by
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Take.lean:46]:
```suggestion
theorem take_eq_init {α : Fin (n + 1) → Sort*} (v : (i : Fin (n + 1)) → α i) :
    take n n.le_succ v = init v := by
```
seems like a more discoverable name

File: Mathlib/Data/Fin/Tuple/Take.lean [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  ext i
+  simp only [Nat.succ_eq_add_one, take, init]
+  congr
+
+@[simp]
+theorem take_all (v : (i : Fin n) → α i) : take n (le_refl n) v = v := by
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Take.lean:52]:
```suggestion
theorem take_eq_self (v : (i : Fin n) → α i) : take n (le_refl n) v = v := by
```

File: Mathlib/Data/Fin/Tuple/Take.lean [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  congr
+
+theorem take_repeat {α : Type*} {n' : ℕ} (m : ℕ) (h : m ≤ n) (a : Fin n' → α) :
+    take (m * n') (Nat.mul_le_mul_right n' h) (Fin.repeat n a) = Fin.repeat m a := by
+  ext i
+  simp only [take, repeat_apply, modNat, coe_castLE]
+
+/-- Taking `m + 1` elements is equal to taking `m` elements and adding the `(m + 1)`th one. -/
+theorem take_succ_eq_snoc (m : ℕ) (h : m < n) (v : (i : Fin n) → α i) :
+    take m.succ h v = snoc (take m (le_of_lt h) v) (v ⟨m, h⟩) := by
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Take.lean:77]:
```suggestion
    take m.succ h v = snoc (take m h.le v) (v ⟨m, h⟩) := by
```

File: Mathlib/Data/Fin/Tuple/Take.lean [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+  · simp only [h', ↓reduceDIte, subNat, castLE, cast, eqRec_eq_cast]
+
+/-- Version of `take_addCases_right` that specializes `addCases` to `append`. -/
+theorem take_append_right {n' : ℕ} {α : Sort*} (m : ℕ) (h : m ≤ n') (u : (i : Fin n) → α)
+    (v : (i : Fin n') → α) : take (n + m) (Nat.add_le_add_left h n) (append u v)
+        = append u (take m h v) :=
+  take_addCases_right m h _ _
+
+/-- `Fin.take` intertwines with `List.take` via `List.ofFn`. -/
+theorem take_eq_take_list_ofFn {α : Type*} {m : ℕ} (h : m ≤ n) (v : Fin n → α) :
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Take.lean:145]:
```suggestion
theorem ofFn_take_eq_take_ofFn {α : Type*} {m : ℕ} (h : m ≤ n) (v : Fin n → α) :
```

File: Mathlib/Data/Fin/Tuple/Take.lean [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    List.ofFn (take m h v) = (List.ofFn v).take m :=
+  List.ext_get (by simp [h]) (fun n h1 h2 => by simp)
+
+/-- Alternative version of `take_eq_take_list_ofFn` with `l : List α` instead of `v : Fin n → α`. -/
+theorem take_eq_take_list_ofFn' {α : Type*} {m : ℕ} (l : List α) (h : m ≤ l.length) :
+    List.ofFn (take m h l.get) = l.take m :=
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Take.lean:151]:
```suggestion
theorem ofFn_take_get {α : Type*} {m : ℕ} (l : List α) (h : m ≤ l.length) :
    List.ofFn (take m h l.get) = l.take m :=
```

File: Mathlib/Data/Fin/Tuple/Take.lean [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+  List.ext_get (by simp [h]) (fun n h1 h2 => by simp)
+
+/-- `Fin.take` intertwines with `List.take` via `List.get`. -/
+theorem take_eq_take_list_get {α : Type*} {m : ℕ} (l : List α) (h : m ≤ l.length) :
+    (l.take m).get = (take m h l.get) ∘ Fin.cast (List.length_take_of_le h) := by
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Take.lean:156]:
```suggestion
theorem get_take_eq_take_get_comp_cast {α : Type*} {m : ℕ} (l : List α) (h : m ≤ l.length) :
    (l.take m).get = take m h l.get ∘ Fin.cast (List.length_take_of_le h) := by
```

File: Mathlib/Data/Fin/Tuple/Take.lean [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+  ext i
+  simp only [List.get_eq_getElem, List.getElem_take, comp_apply, take_apply, coe_castLE, coe_cast]
+
+/-- Alternative version of `take_eq_take_list_get` with `v : Fin n → α` instead of `l : List α`. -/
+theorem take_eq_take_list_get' {α : Type*} {m : ℕ} (v : Fin n → α) (h : m ≤ n) :
+    ((List.ofFn v).take m).get = (take m h v) ∘ Fin.cast (by simp [h]) := by
```
Comment (YaelDillies) [inline Mathlib/Data/Fin/Tuple/Take.lean:162]:
```suggestion
theorem get_take_ofFn_eq_take_comp_cast {α : Type*} {m : ℕ} (v : Fin n → α) (h : m ≤ n) :
    ((List.ofFn v).take m).get = take m h v ∘ Fin.cast (by simp [h]) := by
```

**Review by YaelDillies — COMMENTED** [review]
Thanks for your patience! The end is in sight.

**Review by YaelDillies — APPROVED** [review]
I'm appy with this. What about you, @kim-em ?

maintainer merge

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**